### PR TITLE
MAINT: signal: replace distutils templating with tempita

### PR DESCRIPTION
**Pull request by ev-br**
#### Reference issue

gh-13930

#### What does this implement/fix?

Proof-of-concept / RFC: replace .c.src templating with tempita templating.

#### Additional information

This is only  a proof of concept. 

- [x] there is more distutils templating in `correlate_nd.c.src`
- [x] tempita invocation should be made more robust. The incantation in `signal/setup.py` is taken from `tools/cythonize.py`
- [x] if we go this way, do we keep autogenerated files under source control? In the current state of this PR, `lfilter.c` *is* under source control because 

```
$ git log scipy/signal/lfilter.c
commit aa1882ef55206d746bb3f840a1b33f886547e54b
Author: David Cournapeau <cournape@gmail.com>
Date:   Mon Apr 20 08:41:45 2009 +0000

    rename lfilter.c to a templated file.
``

### CODE REVIEW

**Comment by ilayn** [comment]
Nice! Better keep the generated files out. They are already out for other instances of tempita usage with Cython; pyx.in files. Also otherwise it pollutes the git folder when built in place.

File: scipy/signal/setup.py [hunk 1]
```diff
@@ -1,5 +1,28 @@
...
+import sys
+
+def process_tempita(fromfile):
+    try:
+        try:
+            from Cython import Tempita as tempita
+        except ImportError:
+            import tempita
+    except ImportError:
+        raise Exception("Building requires tempita")
```
Comment (ilayn) [inline scipy/signal/setup.py:12]: Should we rely on `tools/npy_tempita` from NumPy instead?

File: scipy/signal/setup.py [hunk 2]
```diff
@@ -1,5 +1,28 @@
 from scipy._build_utils import numpy_nodepr_api
 import os
+import sys
+
+def process_tempita(fromfile):
```
Comment (rgommers) [inline scipy/signal/setup.py:5]: I think we should put this in `scipy._build_tools` or `tools/`, and vendor `tempita` rather than do the try-except dance.

**Review by rgommers — COMMENTED** [review]
Thanks @ev-br! Looks good overall. Besides moving the tempita processing function to a standalone tool that can be invoked from the command line, I think this looks good.

I agree with @ilayn about not committing the generated sources.

File: scipy/signal/setup.py (hunk 1 above)
Comment (rgommers) [reply scipy/signal/setup.py:12]: Comments crossed. We should vendor that, since `npy_tempita` does not get installed.

Comment (ilayn) [reply scipy/signal/setup.py:12]: ah yes. forgot about that.

File: scipy/signal/setup.py (hunk 2 above)
Comment (ev-br) [reply scipy/signal/setup.py]: Moved it all to _build_utils for now. Maybe push off vendoring until we really need it --- I rather strongly suspect this try-except dance is all dead code as long as cython bundles tempita.

**Comment by ev-br** [comment]
OK, converted all *c.src templates, this PR is ready for review. 
Please do review the C part --- the stuff is reasonably error-prone if very simple.

The comment about keeping *.c files under source control turns out to be moot, `signal/lfilter.c` and `signal/correlate_nd.c` are listed in .gitignore already. 

Re making tempita support more robust, maybe best keep it simple in this PR and polish as needed. I suspect what's here now is good enough for the next release, and all build stuff will need tweaking anyway when we move over to Meson.

**Comment by ev-br** [comment]
CI failure is baffling. Does it expect me to add typing stubs for cython? 

```
scipy/_build_utils/tempita.py:4: error: Skipping analyzing 'Cython': found module but no type hints or library stubs  [import]
scipy/_build_utils/tempita.py:4: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 680 source files)
Error: Process completed with exit code 1.
```

**Comment by rgommers** [comment]
> CI failure is baffling. Does it expect me to add typing stubs for cython?

You can just add `ignore_errors = True` for this file in `mypy.ini`.

File: scipy/_build_utils/tempita.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import sys
+import os
+
+from Cython import Tempita as tempita
+# XXX: If this import ever fails (does it really?), vendor either
```
Comment (rgommers) [inline scipy/_build_utils/tempita.py:5]: I don't think it ever fails anymore now that we do require Cython as a build dependency (wasn't always the case). It's just that it's not public API so Cython devs may decide to remove it at some point. But we can deal with it then.

**Review by rgommers — APPROVED** [review]
LGTM, in it goes. Thanks @ev-br!

**Pull request merged**